### ISSUE TITLE
LaTeX definitions in header-includes escaped or deleted

### ISSUE THREAD

**Issue by zeeMonkeez** [jgm/pandoc#3807]
Using pandoc compiled on macOS with home brew, using commit 3267af2.
Some LaTeX commands included in the `header-includes` field are now either escaped or deleted outright:

Source:
```
---
title: "LaTeX newcommands fail"
header-includes:
- \usepackage{pdflscape}
- \newcommand{\testCom}{testCom}
- \newcommand\testCom1{testCom1}
- \newcommand\benv1{\begin{env1}}
- \newcommand\eenv1{\end{env1}}
- \newcommand{\benv}{\begin{env}}
- \newcommand{\eenv}{\end{env}}
---

test
```

Produces 
```
\usepackage{pdflscape}

\textbackslash{}newcommand\testCom1\{testCom1\}
\textbackslash{}newcommand\benv1\{\textbackslash{}begin\{env1\}\}
\textbackslash{}newcommand\eenv1\{\textbackslash{}end\{env1\}\}



\title{LaTeX newcommands fail}
```

using `pandoc latexdef.md -o latexdef.tex -s --latex-engine=xelatex`. This may have been introduced by 0feb7504b1c68cef76b30ea9987e2eae3101714c, it was working in c505de5 (at least the definitions in braces).

**Comment by jgm** [comment]
If I'm not mistaken, you can't define a command named `\benv1`, because `\benv1` is not a control sequence.  See https://tex.stackexchange.com/questions/242053/what-is-a-control-sequence

You'll see this if you run pdflatex on a file containing
```
\newcommand{\ok1}{OK one}

\ok1 and so on
```

What you're doing may have been working in older versions of pandoc; but it broke because pandoc has become more accurate, not less accurate, in parsing LaTeX.  It never would have worked with LaTeX, as far as I can see.

Feel free to comment here if I'm confused about this.

**Comment by zeeMonkeez** [comment]
OK, maybe there was the misconception on my side about control sequences. The example above still highlights the issue: `\testCom`, `\benv`, and `\eenv` are stripped completely:

```
header-includes:
- \usepackage{pdflscape}
- \newcommand{\testCom}{testCom}
- \newcommand\testComA{testComA}
- \newcommand\benvA{\begin{envA}}
- \newcommand\eenvA{\end{envA}}
- \newcommand{\benv}{\begin{env}}
- \newcommand{\eenv}{\end{env}}
```

yields

```
\usepackage{pdflscape}







```

**Comment by jgm** [comment]
The Markdown reader interprets and applies the newcommand macros itself, rather than passing them on to the output.  If you don't want this (e.g. because you're targeting latex), you can do `-f markdown-latex_macros -t latex -s`.  At least that's how it's now supposed to work.  But I can see that there's still a problem, since now I get:
```
\usepackage{pdflscape}
\newcommand{\testCom}{testCom}
\newcommand

\testComA{testComA}
\newcommand

\benvA{\begin{envA}}
\newcommand

\eenvA{\end{envA}}
\newcommand{\benv}{\begin{env}}
\newcommand{\eenv}{\end{env}}
```

So, reopening.

**Comment by jgm** [comment]
The problem now is that we're not properly reading `\newcommand` as raw LaTeX when the first argument is n't in braces.

**Comment by zeeMonkeez** [comment]
OK, I guess it's not intuitive to me why without specifying `-f markdown-latex_macros` some things get filtered (`\newcommand`) and others don't (`\usepackage`).

The way I'd expect things to work is: If going to (or through) LaTeX, pass on macros by default. If not going through LaTeX, ignore. To evaluate macros for non-LaTeX destinations, specify `-f markdown-latex_macros` (this option should perhaps do nothing if going through LaTeX).

But then again, I know nothing about the underlying structures...

Using `\newcommand` without braces is not something I usually do, but I saw it in a comment as supported behaviour.

**Comment by zeeMonkeez** [comment]
BTW, the minimal example

```
---
title: "LaTeX newcommands fail"
header-includes:
- \newcommand{\testCom}{testCom}
---

\testCom 

```

produces an empty document (up to whitespace) when translated with `pandoc latexdef.md -o latexdef.html -f markdown-latex_macros -t html` which I didn't expect from your description and the documentation. (I'd expect `testCom`)

**Comment by jgm** [comment]
+++ Jonas Zimmermann [Jul 23 17 09:31 ]:
>   OK, I guess it's not intuitive to me why without specifying -f
>   markdown-latex_macros some things get filtered (\newcommand) and others
>   don't (\usepackage).

It's not just a matter of filtering.  Pandoc is actually
applying the macros, replacing your defined commands
with their macro equivalents.

If macros are being applied by pandoc's markdown reader,
then there's no need to pass them through as raw LaTeX.

Though I suppose (in most ordinary cases) it wouldn't hurt.
Maybe we could change to passing them through AND resolving
them.

>   The way I'd expect things to work is: If going to (or through) LaTeX,
>   pass on macros by default. If not going through LaTeX, ignore. To
>   evaluate macros for non-LaTeX destinations, specify -f
>   markdown-latex_macros (this option should perhaps do nothing if going
>   through LaTeX).

Previously pandoc did work this way; macros wouldn't be
resolved, and macro definitions would be passed through,
if the output format was LaTeX.  The new system is less
ad hoc and more compositional.  In general the behavior
of the parser should not depend on which output format
is selected.

**Comment by jgm** [comment]
+++ Jonas Zimmermann [Jul 23 17 09:35 ]:
>   BTW, the minimal example
>---
>title: "LaTeX newcommands fail"
>header-includes:
>- \newcommand{\testCom}{testCom}
>---
>
>\testCom
>
>
>   produces an empty document (up to whitespace) when translated with
>   pandoc latexdef.md -o latexdef.html -f markdown-latex_macros -t html
>   which I didn't expect from your description and the documentation. (I'd
>   expect testCom)

Thanks, this looks like another problem, probably a regression
due to the recent major rewrite of the LaTeX reader.

Macros are getting resolved when input is latex, but not for
markdown.

**Comment by zeeMonkeez** [comment]
> It's not just a matter of filtering.  Pandoc is actually
applying the macros, replacing your defined commands
with their macro equivalents.

I understand that this is the behaviour _with_ `-f
   markdown-latex_macros` specified. Otherwise – as I wrote – some LaTeX macros are replaced by whitespace and others are left as is. And it seems there is some context dependency: compare the result of calling `pandoc latexdef.md -o latexdef.tex -s` when `latexdef.md` contains either this:

```
\usepackage{y}
\newcommand{\testComA}{testComA}
\testComA
```

or this:

```
\newcommand{\testComA}{testComA}
\testComA
```

> If macros are being applied by pandoc's markdown reader,
then there's no need to pass them through as raw LaTeX.

This makes sense, although it is not applied consistently. I would expect the macro substitution then also to happen if the output format is LaTeX, for consistency, because as you write:

>  In general the behavior
of the parser should not depend on which output format
is selected.

But this may generate some side effects (e.g. if source is LaTeX and `\newcommand` itself is redefined).

**Comment by jgm** [comment]
First, what I said was wrong.  The Markdown reader (with the default `latex_macros` extension) only resolves macros in math contexts.

Second, I've changed things so that the macro definitions will remain as raw LaTeX even with `+latex_macros` -- this is especially important because macros aren't resolved outside of math contexts.

Third, I've fixed some bugs, but not yet all of them. More later.

**Comment by jgm** [comment]
OK, I think these issues are now fixed, as of e7876d43203a4672e9d5bb10f91d1530272abc7b.
